[OFFLOAD] Add support for indexed per-thread containers

adurang opened the pull request:
Split from #158900 it adds a PerThreadContainer that can use STL-like indexed containers based on a slightly refactored PerThreadTable.

adurang committed:
[OFFLOAD] Add support for indexed per-thread containers

Split from #158900 it adds a PerThreadContainer that can use STL-like
indexed containers based on a slightly refactored PerThreadTable.

jhuber6 committed:
Merge branch 'main' into extend_ptt

jhuber6 commented on offload/include/PerThreadTable.h:
```diff
@@ -16,17 +16,93 @@
+  };
+
+  std::mutex Mutex;
+  std::list<std::shared_ptr<PerThreadData>> ThreadDataList;
```
Sorry if I already asked this, but why do we need stable iterators here? It's already containing a shared pointer so the data should be stable unless we're saving iterators and returning them.

jhuber6 commented on offload/include/PerThreadTable.h:
```diff
@@ -16,17 +16,93 @@
+  PerThread &operator=(PerThread &&) = delete;
+  ~PerThread() {
+    std::lock_guard<std::mutex> Lock(Mutex);
+    ThreadDataList.clear();
```
I've never been a fan of this kind of implicit synchronization. Don't we have some helpers in `libomptarget` for accessing shared resources? Otherwise I'd assume we can just let the default destructor do its thing.

adurang commented on offload/include/PerThreadTable.h:
```diff
@@ -16,17 +16,93 @@
 template <typename ContainerType, typename ObjectType> struct PerThreadTable {
   using iterator = typename ContainerType::iterator;
 
+  template <typename, typename = std::void_t<>>
```
This is to detect the properties of the underlying container so the template can adjust the code a bit. There's a good overview of how this works here: https://spectre-code.org/sfinae.html

I had a much simpler (and human readable version) using concepts but those are C++20 and we don't seem to be using that yet.

adurang committed:
Change container

adurang committed:
Add deinit method

adurang committed:
renaming variables, remove some locks

adurang committed:
format

adurang committed:
more renaming

adurang committed:
forgot one mutex

adurang committed:
missed one container

adurang commented:
@jhuber6 I fixed the containers, removed most locks and made some cleanup. I also added a new routine that handles deinit with Errors. 

About the locking, all the current cases require the fine-grain locking so I'm not sure if it's worth it to add the option I mentioned at this point (but we could do that later).

jhuber6 commented on offload/include/PerThreadTable.h:
```diff
@@ -100,15 +176,101 @@ template <typename ContainerType, typename ObjectType> struct PerThreadTable {
   }
 
   template <class F> void clear(F f) {
```
Can we give this callback a more descriptive name?

adurang committed:
remove auto

adurang committed:
missed one

adurang committed:
Some renaming

adurang committed:
Add asserts

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -14,26 +14,106 @@
+  PerThread &operator=(PerThread &&) = delete;
+  ~PerThread() {
+    assert(Mutex.try_lock() &&
+           "Cannot be deleted while other threads are adding entries");
```
It's undefined behavior to destroy the mutex while it's locked:
> The behavior of a program is undefined if a mutex is destroyed while still owned by any threads, or a thread terminates while owning a mutex.

Given that there is no function to test the mutex, you may use:

```cpp
assert(M.try_lock() && (M.unlock(), true) && "Cannot ...");
```

adurang committed:
Fix asserts to not be undefined

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -14,26 +14,106 @@
 // Using an STL container (such as std::vector) indexed by thread ID has
 // too many race conditions issues so we store each thread entry into a
 // thread_local variable.
 // T is the container type used to store the objects, e.g., std::vector,
```
Not sure if it's the best place to update this, but this comment is out-of-date

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -91,24 +177,114 @@ template <typename ContainerType, typename ObjectType> struct PerThreadTable {
+
+template <typename ContainerType, size_t reserveSize = 0>
+struct PerThreadContainer
+    : public PerThreadTable<ContainerType,
```
I'm a bit confused. What's the difference between PerThreadTable and PerThreadContainer?

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -14,26 +14,106 @@
+      : std::true_type {};
+
   struct PerThreadData {
     size_t NElements = 0;
```
Why is the `NElements` needed? Can't access the container.size()?

kevinsala replied:
Okay. Can we rename it to `Size`?

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -91,24 +177,114 @@ template <typename ContainerType, typename ObjectType> struct PerThreadTable {
-      ThData->ThEntry->clear(f);
-      ThData->NElements = 0;
+      if constexpr (has_clearAll<ContainerType>::value) {
+        ThreadData->ThreadEntry->clearAll(ClearFunc);
```
What's `clearAll`? I guess that's not part of any STL container, and I don't see any container in the L0Plugin defining that method.

adurang commented on offload/include/PerThreadTable.h:
```diff
@@ -91,24 +177,114 @@ template <typename ContainerType, typename ObjectType> struct PerThreadTable {
+  struct has_reserve : std::false_type {};
+  template <typename T>
+  struct has_reserve<T, std::void_t<decltype(std::declval<T>().reserve(1))>>
+      : std::true_type {};
```
I've been trying unsuccesfully because moving that "code" from the template arguments into a base classe breaks SFINAE. And every example i've found does the same, so not sure it's possible.

At best we could do a macro that would define the template, something like this:
```
#define` DEFINE_CONCEPT_CHECK(Name, Cond) \
  template <typename, typename = std::void_t<>> \
  struct Name : std::false_type {}; \
  template <typename T> \
  struct Name<T, std::void_t<Cond>> : std::true_type {};

#define DEFINE_TYPE_CHECK(Name, Type) \
  DEFINE_CONCEPT_CHECK(Name, typename T::Type)
#define DEFINE_MEMBER_CHECK(Name, Func) \
  DEFINE_CONCEPT_CHECK(Name, decltype(std::declval<T>().Func()))

DEFINE_TYPE_CHECK(has_iterator, iterator)
DEFINE_MEMBER_CHECK(has_clear, clear())
DEFINE_MEMBER_CHECK(has_clearAll, clearAll(1))

```

Is this better?

kevinsala replied:
I think you can use something like this:

```cpp
#include <type_traits>
#include <vector>

template <typename, template <typename> class, typename = std::void_t<>>        
struct has : std::false_type {};                                                
                                                                                
template <typename T, template <typename> class Op>                             
struct has<T, Op, std::void_t<Op<T>>> : std::true_type {};                      
                                                                                
template <typename T>                                                           
using reserve = decltype(std::declval<T>().reserve(1));                        
                                                                                
int main() {                                                                    
    if constexpr (has<std::vector<int>, reserve>::value) {                      
        printf("has reserve\n");                                       
    } else {                                                                    
        printf("hasn't reserve\n");                                    
    }                                                                           
}
```

Then you will only need a `template <typename T> using <...> = decltype(...)` for each function or property to check.

adurang committed:
minor refactoring

adurang committed:
Refactor container concepts checks

adurang committed:
remove clearAll

adurang committed:
Simplify PerThreadData

adurang committed:
Address comments

jhuber6 commented on offload/include/PerThreadTable.h:
```diff
@@ -14,101 +14,256 @@
+    if (!ThreadData) {
+      ThreadData = std::make_shared<ObjectType>();
+      std::lock_guard<std::mutex> Lock(Mutex);
+      ThreadDataList.push_back(ThreadData);
```
Could we do an emplace back here instead to create the smart pointer directly?

kevinsala reviewed (approved):
I'm still not convinced of the `table`/`container` names, and the extra field `NElements`/`size`. But I think it can be fixed in another PR later.

LGTM now.

jhuber6 reviewed (approved):
Seems fine overall, some nits

kevinsala commented on offload/include/PerThreadTable.h:
```diff
@@ -14,101 +14,256 @@
+  ObjectType &get(IndexType Index) {
+    ContainerType &Entry = this->getThreadEntry();
+
+    // specialized code for vector-like containers
```
Uppercase

adurang committed:
fix comments
